packages/tizen-remote/lib/index.js

```diff
@@ -0,0 +1,764 @@
+   */
+  get isDisconnected() {
+    return Boolean(this.#ws?.readyState === WebSocket.CLOSED);
+  }
+
+  /**
+   * `true` if we've issued a manual disconnection
+   */
+  get isDisconnecting() {
+    return Boolean(this.#ws?.readyState === WebSocket.CLOSING);
+  }
+
+  /**
+   * `true` if we are currently attempting to connect to the server.
+   */
+  get isConnecting() {
+    return Boolean(this.#ws?.readyState === WebSocket.CONNECTING);
+  }
+
+  /**
+   *
+   * @param {import('ws').RawData} data
+   * @param {boolean} isBinary
+   */
+  #debugListener(data, isBinary) {
+    if (isBinary) {
+      this.#debug('Received binary message: %o', data);
+    } else {
+      try {
+        const resData = JSON.parse(data.toString());
+        this.#debug('Received message: %o', resData);
+      } catch {}
+    }
+  }
+
+  /**
+   * This is the definition of a listener for the {@linkcode WsEvent#CLOSE} event
+   * which handles unexpected disconnections.
+   *
+   * When listening, this method will be bound to this instance.
+   *
+   * Note: `reason` mostly seems to be an empty `Buffer`. If someone sees it actually
+   * contain any information, please [tell us](https://github.com/headspinio/appium-tizen-tv-driver/issues/new).
+   * @param {number} code - Disconnection code; see {@linkcode BadCode}
+   * @param {Buffer} reason - Disconnection reason
+   */
+  async #closeListener(code, reason) {
+    this.emit(Event.DISCONNECT, {code, reason});
+    if (isKnownBadCode(code)) {
+      this.#debug('Received "%s" (%d)', BadCode[code], code);
+      if (this.#autoReconnect) {
+        try {
+          await this.connect();
+        } catch (err) {
+          this.emit(Event.ERROR, /** @type {Error} */ (err));
+        }
+      } else {
+        this.emit(
+          Event.ERROR,
+          new Error(`Socket disconnected; ${BadCode[code]}: ${reason || '(no reason provided)'}`)
+        );
+      }
+    }
+  }
+
+  /**
+   * Disconnect from the WS server.
+   *
+   * If not connected, this method does nothing.  If disconnection
+   * already in progress, the `Promise` will fulfill upon disconnection.
+   * @returns {Promise<void>}
+   */
+  async disconnect() {
+    return await new Promise((resolve, reject) => {
+      // nothing to do!
+      if (!this.#ws || this.isDisconnected) {
+        resolve();
+        return;
+      }
+
+      // disconnecting already in progress; easiest to just
+      // wait for our own event.
+      if (this.isDisconnecting) {
+        this.once(Event.DISCONNECT, () => {
+          resolve();
+        });
+        return;
+      }
+
+      this.emit(Event.DISCONNECTING);
+
+      this.#ws
+        .once(WsEvent.CLOSE, () => {
+          this.#debug('Closed connection to server %s', this.#url);
+          this.emit(Event.DISCONNECT);
+          resolve();
+        })
+        .once(WsEvent.ERROR, (err) => {
+          // XXX: rejection here might not be appropriate
+          reject(err);
```

Comment:
my feeling is it is appropriate. the consumer can always wrap in exception handling if they don't care about errors during disconnect?

Reply:
what confuses me is the interplay between EE's and promises.  maybe emitting `error` is something streams should do, and the fact that non-stream EE's emit `error` is just a historical artifact.  so the alternative here, in my mind (which I did not write in the comment), was emitting `error` instead.  I think we should just keep it as-is.